src/hnsw/build.c

```diff
@@ -166,6 +221,11 @@ static void InitBuildState(HnswBuildState *buildstate, Relation heap, Relation i
 
     buildstate->dimensions = GetHnswIndexDimensions(index);
 
+    // If a dimension wasn't specified try to infer it
+    // this is unecessary when the actual size is 3 but in practice this shouldn't be too common
+    if(columnType == REAL_ARRAY || columnType == INT_ARRAY)
+        if(buildstate->dimensions == HNSW_DEFAULT_DIMS)
+            buildstate->dimensions = GetArrayLengthFromHeap(heap, indexInfo, buildstate->dimensions);
     /* Require column to have dimensions to be indexed */
     if(buildstate->dimensions < 1) elog(ERROR, "column does not have dimensions");
```

Comment:
Is this the error we are going to see if the table is empty? Is that case tested?

Reply:
It would be if `DEFAULT_DIMS` were changed from 3 to -1.  Currently a number of tests create indices on empty tables, so I wasn't sure what the expected behavior was. I preserved the default to maximize compatibility with the existing API, so right now this wouldn't generate an error. If it's alright to throw an error when the table being indexed is empty I can change the default per @var77's suggestion and update the relevant tests to make things more consistent

Reply:
It is all right and the expected behaviour. Just throw a helpful message asking the user to provide number of dimensions for the indexed column